task_ner_labels = {
    "bc5cdr": ["Disease", "Chemical"],
    "bc4chemdner": ["Chemical"],
    "bc8": [
        "DiseaseOrPhenotypicFeature",
        "ChemicalEntity",
        "GeneOrGeneProduct",
        "SequenceVariant",
        "OrganismTaxon",
        "CellLine",
    ],
    "anat_em": [
        "Organism_substance",
        "Organ",
        "Cell",
        "Cellular_component",
        "Tissue",
        "Cancer",
        "Multi-tissue_structure",
        "Anatomical_system",
        "Organism_subdivision",
        "Pathological_formation",
        "Immaterial_anatomical_entity",
        "Developing_anatomical_structure",
    ],
    "an_em": [
        "Tissue",
        "Organism_substance",
        "Organ",
        "Anatomical_system",
        "Organism_subdivision",
        "Multi-tissue_structure",
        "Cellular_component",
        "Pathological_formation",
        "Immaterial_anatomical_entity",
        "Cell",
        "Developing_anatomical_structure",
    ],
    "bionlp_st_2011_epi": ["Protein", "Entity"],
    "bionlp_st_2011_ge": ["Protein", "Entity"],
    "bioinfer": [
        "Individual_protein",
        "Protein_family_or_group",
        "Gene/protein/RNA",
        "Gene",
        "DNA_family_or_group",
        "Protein_complex",
    ],
    "bionlp_st_2013_gro": [
        "Protein",
        "Tissue",
        "CellComponent",
        "Eukaryote",
        "Cell",
        "Transcript",
        "Chromosome",
        "ProteinIdentification",
        "DNA",
        "AminoAcid",
        "ProteinDomain",
        "MessengerRNA",
        "Sequence",
        "ProteinSubunit",
        "Enzyme",
        "Holoenzyme",
        "ProteinComplex",
        "Peptide",
        "BindingAssay",
        "CatalyticActivity",
        "MutantProtein",
        "Phenotype",
        "NucleicAcid",
        "TranscriptionCofactor",
        "TranslationFactor",
        "RNA",
        "Chemical",
        "Ligand",
        "ExperimentalMethod",
        "Ion",
        "BindingSiteOfProtein",
        "Gene",
        "OpenReadingFrame",
        "GeneProduct",
        "SequenceHomologyAnalysis",
        "InorganicChemical",
        "Substrate",
        "TranscriptionFactor",
        "TranscriptionFactorBindingSiteOfDNA",
        "bZIP",
        "TranscriptionFactorComplex",
        "Enhancer",
        "HomeoboxTF",
        "RegulatoryDNARegion",
        "BasicDomain",
        "Locus",
        "Nucleotide",
        "MolecularEntity",
        "ExpressionProfiling",
        "TranscriptionRegulator",
        "OrganicChemical",
        "MolecularStructure",
        "ComplexMolecularEntity",
        "Conformation",
        "ThreeDimensionalMolecularStructure",
        "Prokaryote",
        "LivingEntity",
        "Exon",
        "Intron",
        "ZincCoordinatingDomain",
        "AntisenseRNA",
        "ProteinCodingRegion",
        "tRNA",
        "Virus",
        "Nucleus",
        "TranscriptionActivator",
        "Promoter",
        "Lipid",
        "TATAbox",
        "DNABindingDomainOfProtein",
        "DNAFragment",
        "TranscriptionFactorActivity",
        "Microorganism",
        "Function",
        "SecondMessenger",
        "TranscriptionActivatorActivity",
        "HMGTF",
        "HMG",
        "MutatedProtein",
        "MolecularFunction",
        "Ribosome",
        "TranscriptionRepressor",
        "ZincCoordinatingDomainTF",
        "ReporterGene",
        "TranscriptionRepressorActivity",
        "Histone",
        "FusionOfGeneWithReporterGene",
        "ForkheadWingedHelixTF",
        "ProteinBindingSiteOfProtein",
        "TranscriptionCorepressor",
        "ProteinBindingSiteOfDNA",
        "UpstreamRegulatorySequence",
        "DNARegion",
        "ChromosomalDNA",
        "DNAMolecule",
        "Stress",
        "Attenuator",
        "DNABindingSite",
        "EukaryoticCell",
        "SmallInterferingRNA",
        "CpGIsland",
        "ReporterGeneConstruction",
        "AP2EREBPRelatedDomain",
        "PositiveTranscriptionRegulator",
        "bZIPTF",
        "PrimaryStructure",
        "OxidativeStress",
        "ForkheadWingedHelix",
        "HomeoBox",
        "RNAPolymerase",
        "FusionGene",
        "DoubleStrandDNA",
        "bHLH",
        "bHLHTF",
        "RibosomalRNA",
        "RuntLikeDomain",
        "TranscriptionCoactivator",
        "NuclearReceptor",
        "FusionProtein",
        "PhysicalContinuant",
        "Heterochromatin",
        "Chromatin",
        "PhysicalContact",
        "Operon",
        "Polymerase",
        "ComplexOfProteinAndDNA",
        "ProteinCodingDNARegion",
        "RNAPolymeraseII",
    ],
    "bionlp_st_2019_bb": ["Habitat", "Microorganism", "Phenotype"],
    "bionlp_shared_task_2009": ["Protein", "Entity"],
    "bionlp_st_2013_pc": [
        "Gene_or_gene_product",
        "Complex",
        "Cellular_component",
        "Simple_chemical",
    ],
    "bionlp_st_2013_ge": ["Protein", "Anaphora", "Entity"],
    "bionlp_st_2011_rel": ["Protein", "Entity"],
    "bionlp_st_2011_id": [
        "Organism",
        "Protein",
        "Two-component-system",
        "Chemical",
        "Regulon-operon",
        "Entity",
    ],
    "bionlp_st_2013_cg": [
        "Cell",
        "Organism",
        "Gene_or_gene_product",
        "Organism_substance",
        "Cellular_component",
        "Tissue",
        "Simple_chemical",
        "Cancer",
        "Organ",
        "Multi-tissue_structure",
        "Amino_acid",
        "Protein_domain_or_region",
        "Anatomical_system",
        "Pathological_formation",
        "Organism_subdivision",
        "Immaterial_anatomical_entity",
        "DNA_domain_or_region",
        "Developing_anatomical_structure",
    ],
    "hprd50": ["protein"],
    "genetag": ["NEWGENE"],
    "genia_ptm_event_corpus": ["Protein", "Entity"],
    "jnlpba": ["1", "2", "0", "9", "10", "7", "8", "5", "6", "3", "4"],
    "genia_term_corpus": [
        "other_name",
        "DNA_domain_or_region",
        "protein_molecule",
        "protein_family_or_group",
        "cell_type",
        "inorganic",
        "protein_complex",
        "virus",
        "DNA_N/A",
        "DNA_family_or_group",
        "cell_line",
        "other_organic_compound",
        "multi_cell",
        "protein_domain_or_region",
        "atom",
        "(AND DNA_domain_or_region DNA_domain_or_region)",
        "",
        "lipid",
        "DNA_molecule",
        "RNA_molecule",
        "RNA_family_or_group",
        "(AND DNA_family_or_group DNA_family_or_group)",
        "(AND cell_type cell_type cell_type)",
        "(AND other_name other_name)",
        "(AND RNA_molecule RNA_molecule)",
        "(AND cell_type cell_type)",
        "protein_substructure",
        "amino_acid_monomer",
        "(AND protein_subunit protein_subunit)",
        "protein_subunit",
        "(AND other_artificial_source other_artificial_source)",
        "(AND cell_line cell_line)",
        "cell_component",
        "protein_N/A",
        "(BUT_NOT other_name other_name)",
        "tissue",
        "(AND other_name other_name other_name other_name)",
        "(AND protein_family_or_group protein_family_or_group)",
        "nucleotide",
        "peptide",
        "body_part",
        "(AND protein_molecule protein_molecule protein_molecule protein_molecule)",
        "(AND protein_molecule protein_molecule)",
        "polynucleotide",
        "DNA_substructure",
        "(AND cell_line cell_line cell_line cell_line cell_line cell_line)",
        "(AND cell_line cell_line cell_line)",
        "(OR other_name other_name)",
        "(OR protein_family_or_group protein_family_or_group)",
        "(AS_WELL_AS other_name other_name)",
        "other_artificial_source",
        "(AND protein_N/A protein_N/A)",
        "(AND DNA_molecule DNA_molecule)",
        "(OR DNA_domain_or_region DNA_domain_or_region)",
        "RNA_domain_or_region",
        "(AND protein_molecule protein_molecule protein_molecule)",
        "(OR protein_molecule protein_molecule protein_molecule)",
        "(AND/OR DNA_domain_or_region DNA_domain_or_region)",
        "(AND protein_molecule protein_molecule protein_molecule protein_molecule protein_molecule protein_molecule)",
        "(AND DNA_domain_or_region DNA_domain_or_region DNA_domain_or_region DNA_domain_or_region)",
        "(AND protein_domain_or_region protein_domain_or_region)",
        "(AND multi_cell multi_cell)",
        "(AND cell_type cell_type cell_type cell_type)",
        "(AND other_name other_name other_name)",
        "(AND DNA_domain_or_region DNA_domain_or_region DNA_domain_or_region)",
        "(AND other_organic_compound other_organic_compound)",
        "mono_cell",
        "(AND lipid lipid)",
        "(OR DNA_molecule DNA_molecule DNA_molecule DNA_molecule DNA_molecule)",
        "(BUT_NOT protein_molecule protein_molecule)",
        "(AND protein_complex protein_complex)",
        "(OR DNA_family_or_group DNA_family_or_group)",
        "(AND_NOT cell_line cell_line)",
        "RNA_N/A",
        "carbohydrate",
        "(AND tissue tissue)",
        "(AND protein_molecule protein_molecule protein_molecule protein_molecule protein_molecule)",
        "(AND other_name other_name other_name other_name other_name)",
        "(OR other_name other_name other_name)",
        "(OR protein_domain_or_region protein_domain_or_region)",
        "(OR cell_line cell_line)",
        "(BUT_NOT protein_molecule (OR protein_molecule protein_molecule))",
        "(AND DNA_domain_or_region DNA_domain_or_region DNA_domain_or_region DNA_domain_or_region DNA_domain_or_region DNA_domain_or_region DNA_domain_or_region)",
        "(AND body_part body_part body_part)",
        "(AND carbohydrate carbohydrate)",
        "(AND cell_line cell_line cell_line cell_line)",
        "(OR protein_molecule protein_molecule)",
        "(AND cell_component cell_component cell_component)",
        "(OR virus virus)",
        "(AND amino_acid_monomer amino_acid_monomer amino_acid_monomer)",
        "(AND virus virus)",
        "(AND DNA_domain_or_region DNA_domain_or_region DNA_domain_or_region DNA_domain_or_region DNA_domain_or_region)",
        "(OR protein_family_or_group protein_family_or_group protein_family_or_group protein_family_or_group protein_family_or_group)",
        "(TO other_name other_name)",
        "(NEITHER_NOR RNA_molecule RNA_molecule)",
        "(OR DNA_domain_or_region DNA_domain_or_region DNA_domain_or_region)",
        "(THAN cell_type cell_type)",
        "(OR protein_complex protein_complex)",
        "(OR polynucleotide polynucleotide)",
        "(BUT_NOT cell_line cell_line)",
        "(AND cell_type cell_type cell_type cell_type cell_type cell_type)",
        "(BUT_NOT cell_type cell_type)",
        "(OR other_name other_name other_name other_name)",
        "(OR cell_type cell_type)",
        "(AND RNA_molecule RNA_molecule RNA_molecule RNA_molecule)",
        "(OR other_organic_compound other_organic_compound other_organic_compound)",
        "(AND protein_family_or_group protein_family_or_group protein_family_or_group)",
        "(AND RNA_molecule RNA_molecule RNA_molecule)",
        "(BUT_NOT other_name (OR other_name other_name other_name))",
        "(BUT_NOT RNA_molecule RNA_molecule)",
        "(AND DNA_family_or_group DNA_family_or_group DNA_family_or_group)",
        "(BUT_NOT cell_line (AND cell_line cell_line cell_line))",
        "(OR (AND protein_complex protein_complex) (AND protein_complex protein_complex))",
        "(BUT_NOT (AND cell_type cell_type) (OR cell_type cell_type))",
        "(AND peptide peptide)",
        "(OR cell_line cell_line cell_line)",
        "(AND atom atom)",
        "(AND/OR protein_family_or_group protein_family_or_group)",
        "(AND multi_cell multi_cell multi_cell)",
        "(AND cell_type cell_type cell_type cell_type cell_type)",
        "(AND peptide peptide peptide)",
        "(AND amino_acid_monomer amino_acid_monomer)",
        "(AND body_part body_part)",
        "(OR other_organic_compound other_organic_compound)",
        "(OR multi_cell multi_cell)",
        "(OR protein_subunit protein_subunit)",
        "(AND_NOT cell_type cell_type)",
        "(AND cell_component cell_component)",
        "(AND/OR other_name other_name)",
        "(AS_WELL_AS DNA_domain_or_region DNA_domain_or_region)",
        "(OR RNA_molecule RNA_molecule)",
        "(OR protein_domain_or_region protein_domain_or_region protein_domain_or_region)",
        "(VERSUS protein_subunit protein_subunit)",
        "(AND polynucleotide polynucleotide)",
        "(AND/OR RNA_molecule RNA_molecule)",
        "(OR DNA_molecule DNA_molecule)",
        "(AND/OR cell_type cell_type)",
        "(AND protein_family_or_group protein_family_or_group protein_family_or_group protein_family_or_group)",
        "(OR nucleotide nucleotide nucleotide)",
        "(AND nucleotide nucleotide nucleotide nucleotide nucleotide)",
        "(AND nucleotide nucleotide nucleotide nucleotide)",
        "(BUT_NOT protein_family_or_group protein_family_or_group)",
        "(OR lipid lipid lipid)",
        "RNA_substructure",
        "(BUT_NOT DNA_domain_or_region DNA_domain_or_region)",
        "(TO cell_type cell_type)",
        "(OR tissue tissue)",
        "(AND RNA_molecule RNA_molecule RNA_molecule RNA_molecule RNA_molecule)",
        "(OR DNA_molecule DNA_molecule DNA_molecule)",
        "(AND cell_line (OR cell_line cell_line))",
        "(NOT_ONLY_BUT_ALSO (OR other_name other_name) other_name)",
        "(OR DNA_substructure DNA_substructure)",
        "(OR RNA_family_or_group RNA_family_or_group)",
        "(AND RNA_family_or_group RNA_family_or_group)",
        "(AND (OR other_name other_name) (OR other_name other_name))",
        "(AND (AND other_name other_name) (AND other_name other_name))",
        "(AND protein_domain_or_region protein_domain_or_region protein_domain_or_region protein_domain_or_region)",
        "(AND protein_complex protein_complex protein_complex protein_complex)",
        "(OR inorganic inorganic)",
        "(AND protein_substructure protein_substructure)",
        "(AND RNA_N/A RNA_N/A)",
    ],
    "euadr": [
        "Diseases & Disorders",
        "Chemicals & Drugs",
        "",
        "SNP & Sequence variations",
        "Genes & Molecular Sequences",
    ],
    "medmentions": [
        "T116",
        "T123",
        "T047",
        "T101",
        "T079",
        "T169",
        "T033",
        "T081",
        "T063",
        "T052",
        "T062",
        "T032",
        "T098",
        "T100",
        "T073",
        "T093",
        "T026",
        "T028",
        "T007",
        "T045",
        "T046",
        "T109",
        "T122",
        "T074",
        "T082",
        "T058",
        "T070",
        "T080",
        "T067",
        "T167",
        "T057",
        "T025",
        "T023",
        "T059",
        "T042",
        "T077",
        "T060",
        "T184",
        "T024",
        "T029",
        "T078",
        "T053",
        "T061",
        "T083",
        "T055",
        "T185",
        "UnknownType",
        "T054",
        "T170",
        "T097",
        "T041",
        "T095",
        "T131",
        "T071",
        "T204",
        "T168",
        "T121",
        "T099",
        "T019",
        "T201",
        "T039",
        "T040",
        "T031",
        "T126",
        "T191",
        "T090",
        "T034",
        "T197",
        "T091",
        "T114",
        "T037",
        "T008",
        "T102",
        "T002",
        "T012",
        "T011",
        "T049",
        "T129",
        "T127",
        "T056",
        "T125",
        "T005",
        "T043",
        "T001",
        "T048",
        "T065",
        "T022",
        "T015",
        "T038",
        "T096",
        "T030",
        "T044",
        "T086",
        "T192",
        "T130",
        "T089",
        "T190",
        "T103",
        "T104",
        "T092",
        "T195",
        "T016",
        "T051",
        "T004",
        "T171",
        "T196",
        "T120",
        "T068",
        "T050",
        "T020",
        "T013",
        "T064",
        "T194",
        "T087",
        "T085",
        "T075",
        "T069",
        "T072",
        "T200",
        "T018",
        "T066",
        "T010",
        "T017",
        "T094",
        "T014",
        "T021",
        "T203",
    ],
    "linnaeus": ["species"],
    "gnormplus": ["FamilyName", "Gene", "DomainMotif"],
    "ebm_pico": [
        "Intervention_Pharmacological",
        "Outcome_Physical",
        "Outcome_Other",
        "Participant_Sample-size",
        "Participant_Condition",
        "Intervention_Physical",
        "Intervention_Surgical",
        "Intervention_Control",
        "Outcome_Adverse-effects",
        "Outcome_Mental",
        "Participant_Sex",
        "Outcome_Pain",
        "Intervention_Educational",
        "Participant_Age",
        "Outcome_Mortality",
        "Intervention_Other",
        "Intervention_Psychological",
    ],
    "chebi_nactem": [
        "Protein",
        "Chemical",
        "Biological_Activity",
        "Species",
        "Metabolite",
        "Spectral_Data",
    ],
    "biorelex": [
        "DNA",
        "protein",
        "chemical",
        "protein-isoform",
        "protein-family",
        "peptide",
        "experiment-tag",
        "fusion-protein",
        "protein-domain",
        "protein-complex",
        "cell",
        "gene",
        "reagent",
        "experimental-construct",
        "RNA",
        "protein-motif",
        "organelle",
        "assay",
        "mutation",
        "disease",
        "protein-region",
        "process",
        "drug",
        "tissue",
        "amino-acid",
        "protein-RNA-complex",
        "protein-DNA-complex",
        "parameter",
        "organism",
        "brand",
        "RNA-family",
    ],
    "chemdner": ["Chemical"],
    "chemprot": ["CHEMICAL", "GENE-N", "GENE-Y"],
    "distemist": ["ENFERMEDAD"],
    "ddi_corpus": ["DRUG", "GROUP", "BRAND", "DRUG_N"],
    "ctebmsp": ["CHEM", "PROC", "ANAT", "DISO"],
    "chia": [
        "Condition",
        "Procedure",
        "Qualifier",
        "Temporal",
        "Observation",
        "Scope",
        "Drug",
        "Measurement",
        "Value",
        "Multiplier",
        "Device",
        "Negation",
        "Person",
        "Reference_point",
        "Visit",
        "Mood",
    ],
    "mlee": [
        "Drug_or_compound",
        "Gene_or_gene_product",
        "Cell",
        "Pathological_formation",
        "Tissue",
        "Organism",
        "Organism_substance",
        "Multi-tissue_structure",
        "Organ",
        "Cellular_component",
        "Protein_domain_or_region",
        "Anatomical_system",
        "Developing_anatomical_structure",
        "Immaterial_anatomical_entity",
        "Organism_subdivision",
        "DNA_domain_or_region",
    ],
    "mirna": [
        "Non-Specific_miRNAs",
        "Diseases",
        "Relation_Trigger",
        "Specific_miRNAs",
        "Species",
        "Genes/Proteins",
    ],
    "nlmchem": ["Chemical"],
    "nlm_gene": ["GENERIF", "Gene", "STARGENE", "Domain", "Other"],
    "ncbi_disease": ["DiseaseClass", "SpecificDisease", "CompositeMention", "Modifier"],
    "pdr": ["Disease", "Plant"],
    "osiris": ["gene", "variant"],
    "progene": ["progene_text"],
    "pico_extraction": ["outcome", "intervention", "participant"],
    "tmvar_v2": ["SNP", "DNAMutation", "ProteinMutation"],
    "tmvar_v1": ["ProteinMutation", "DNAMutation", "SNP"],
    "spl_adr_200db": [
        "AdverseReaction",
        "Severity",
        "Factor",
        "DrugClass",
        "Negation",
        "Animal",
    ],
    "seth_corpus": ["Gene", "SNP", "RS"],
    "scai_disease": ["DISEASE", "ADVERSE"],
    "scai_chemical": [
        "IUPAC",
        "TRIVIAL",
        "MODIFIER",
        "FAMILY",
        "PARTIUPAC",
        "",
        "ABBREVIATION",
        "SUM",
        "TRIVIALVAR",
    ],
    "BC5_chem": ["AIO_Chemical"],
    "BC5_dis": ["AIO_Disease"],
    "BioID": ["AIO_CellLine"],
    "NLM_chem": ["AIO_Chemical"],
    "GNormPlus": ["AIO_Gene"],
    "Linnaeus": ["AIO_Species"],
    "ncbi": ["AIO_Disease"],
    "NLMgene": ["AIO_Gene"],
    "Species_800": ["AIO_Species"],
    "tmVar3": ["AIO_Variant"],
    "BC5CDR_TrainDev": ["Chemical", "Disease"],
    "BioID_Train_revision": ["CellLine"],
    "DrugProt_Train": ["Chemical", "Gene"],
    "GNormPlus_Train": ["FamilyName", "Gene", "DomainMotif"],
    "Linnaeus_Train": ["Species"],
    "NCBIdisease_TrainDev": [
        "DiseaseClass",
        "SpecificDisease",
        "CompositeMention",
        "Modifier",
    ],
    "NLMchem_TrainDev": ["Chemical", "NonStandardRef", "OTHER"],
    "NLMgene_Train": [
        "Gene",
        "FamilyName",
        "Cell",
        "DomainMotif",
        "ChromosomeLocation",
    ],
    "SPECIES_Train": ["Species"],
    "tmVar3_Train": [
        "Gene",
        "Species",
        "Disease",
        "DNAMutation",
        "ProteinMutation",
        "OtherMutation",
        "CellLine",
        "AcidChange",
        "SNP",
        "DNAAllele",
        "ProteinAllele",
    ],
}

ner_labels_for_tsne = {
    "bc8": [
        "Disease_BioRED",
        "Chemica_BioRED",
        "Gene_BioRED",
        "Variant_BioRED",
        "Species_BioRED",
        "CellLine_BioRED",
    ],
    "BC5CDR_TrainDev": ["Chemical_BC5CDR", "Disease_BC5CDR"],
    "BioID_Train_revision": ["CellLine_BC5CDR"],
    "DrugProt_Train": ["Chemical_DrugProt", "Gene_DrugProt"],
    "GNormPlus_Train": [
        "FamilyName_GnormPlus",
        "Gene_GnormPlus",
        "DomainMotif_GnormPlus",
    ],
    "Linnaeus_Train": ["Species_Linnaeus"],
    "NCBIdisease_TrainDev": [
        "DiseaseClass_NCBIdisease",
        "SpecificDisease_NCBIdisease",
        "CompositeMention_NCBIdisease",
        "Modifier_NCBIdisease",
    ],
    "NLMchem_TrainDev": ["Chemical_NLMChem", "NonStandardRef_NLMChem", "OTHER_NLMChem"],
    "NLMgene_Train": [
        "Gene_NLMGene",
        "FamilyName_NLMGene",
        "Cell_NLMGene",
        "DomainMotif_NLMGene",
        "ChromosomeLocation_NLMGene",
    ],
    "SPECIES_Train": ["Species_SPECIES"],
    "tmVar3_Train": [
        "Gene_tmVar3",
        "Species_tmVar3",
        "Disease_tmVar3",
        "DNAMutation_tmVar3",
        "ProteinMutation_tmVar3",
        "OtherMutation_tmVar3",
        "CellLine_tmVar3",
        "AcidChange_tmVar3",
        "SNP_tmVar3",
        "DNAAllele_tmVar3",
        "ProteinAllele_tmVar3",
    ],
}


general_label_map = {
    "BC5_chem": {1: 2},
    "BC5_dis": {1: 1},
    "BioID": {1: 6},
    "NLM_chem": {1: 2},
    "GNormPlus": {1: 3},
    "Linnaeus": {1: 5},
    "ncbi": {1: 1},
    "NLMgene": {1: 3},
    "Species_800": {1: 5},
    "tmVar3": {1: 4},
    "bionlp_st_2019_bb": {},
    "euadr": {1: 1, 2: 2, 4: 4, 5: 3},
    "gnormplus": {2: 3},
    "mirna": {2: 1, 5: 5, 6: 3},
    "osiris": {1: 3, 2: 4},
    "tmvar_v2": {1: 4},
    "tmvar_v1": {3: 4},
    "seth_corpus": {1: 3, 2: 4},
    "scai_disease": {1: 1},
}

general_label_map_other = {
    "BC5_chem": {0: 0, 1: 2},
    "BC5_dis": {0: 0, 1: 1},
    "BioID": {0: 0, 1: 6},
    "NLM_chem": {0: 0, 1: 2},
    "GNormPlus": {0: 0, 1: 3},
    "Linnaeus": {0: 0, 1: 5},
    "ncbi": {0: 0, 1: 1},
    "NLMgene": {0: 0, 1: 3},
    "Species_800": {0: 0, 1: 5},
    "tmVar3": {0: 0, 1: 4},
    "bc5cdr": {0: 0, 1: 1, 2: 2},
    "bc4chemdner": {0: 0, 1: 2},
    "anat_em": {0: 0, 1: 5, 2: 3, 3: 6, 4: 6, 9: 5, 10: 3, 11: 2},
    "an_em": {0: 0, 2: 5, 3: 3, 5: 5, 7: 6, 8: 3, 9: 2, 10: 6},
    "bionlp_st_2011_epi": {0: 0, 2: 2},
    "bionlp_st_2011_ge": {0: 0, 2: 2},
    "bioinfer": {0: 0, 2: 3, 3: 3, 4: 3, 5: 3},
    "bionlp_st_2013_gro": {
        0: 0,
        3: 6,
        5: 6,
        13: 4,
        24: 3,
        25: 3,
        27: 2,
        32: 3,
        34: 3,
        35: 4,
        36: 3,
        38: 3,
        39: 3,
        41: 3,
        44: 3,
        48: 2,
        50: 3,
        51: 3,
        53: 2,
        54: 3,
        57: 2,
        60: 3,
        66: 3,
        72: 3,
        73: 5,
        76: 3,
        82: 3,
        83: 3,
        84: 3,
        85: 3,
        87: 3,
        88: 3,
        90: 3,
        92: 4,
        97: 3,
        99: 6,
        102: 3,
        104: 3,
        108: 3,
        111: 3,
        117: 3,
        118: 3,
    },
    "bionlp_st_2019_bb": {0: 0, 2: 5},
    "bionlp_shared_task_2009": {0: 0, 2: 2},
    "bionlp_st_2013_pc": {0: 0, 1: 3, 3: 6, 4: 2},
    "bionlp_st_2013_ge": {0: 0, 2: 3, 3: 2},
    "bionlp_st_2011_rel": {0: 0, 2: 2},
    "bionlp_st_2011_id": {0: 0, 1: 5, 4: 2, 6: 2},
    "bionlp_st_2013_cg": {
        0: 0,
        1: 6,
        2: 5,
        3: 3,
        4: 5,
        5: 6,
        7: 2,
        9: 3,
        12: 3,
        14: 3,
        15: 5,
        16: 2,
        17: 3,
    },
    "hprd50": {0: 0},
    "genetag": {0: 0, 1: 3},
    "genia_ptm_event_corpus": {0: 0, 2: 2},
    "jnlpba": {0: 0},
    "genia_term_corpus": {
        0: 0,
        2: 3,
        4: 3,
        5: 6,
        6: 3,
        10: 3,
        11: 6,
        12: 3,
        13: 6,
        14: 3,
        16: 3,
        21: 3,
        22: 3,
        23: 6,
        26: 6,
        32: 6,
        33: 6,
        38: 3,
        46: 6,
        47: 6,
        48: 3,
        49: 3,
        54: 3,
        55: 3,
        57: 3,
        58: 3,
        60: 3,
        61: 3,
        62: 6,
        63: 6,
        65: 3,
        66: 3,
        67: 6,
        69: 3,
        72: 3,
        73: 6,
        79: 3,
        80: 3,
        81: 6,
        82: 3,
        83: 3,
        86: 6,
        87: 3,
        88: 6,
        89: 3,
        92: 3,
        93: 3,
        95: 3,
        96: 3,
        97: 6,
        98: 3,
        99: 3,
        100: 6,
        101: 6,
        102: 6,
        103: 3,
        104: 6,
        106: 3,
        107: 3,
        109: 3,
        111: 3,
        112: 6,
        113: 3,
        114: 6,
        116: 6,
        118: 3,
        119: 6,
        120: 6,
        124: 3,
        125: 6,
        126: 3,
        127: 6,
        128: 6,
        129: 3,
        130: 3,
        131: 3,
        132: 3,
        135: 3,
        136: 3,
        137: 6,
        138: 3,
        139: 3,
        142: 3,
        143: 3,
        145: 3,
        146: 6,
        147: 3,
        149: 3,
        150: 6,
        151: 3,
        152: 3,
        153: 3,
        154: 3,
        155: 3,
        157: 3,
        159: 3,
    },
    "euadr": {0: 0, 1: 3, 2: 2, 4: 4, 5: 4},
    "medmentions": {0: 0},
    "linnaeus": {0: 0},
    "gnormplus": {0: 0, 2: 3},
    "ebm_pico": {0: 0, 15: 3},
    "chebi_nactem": {0: 0, 2: 2},
    "biorelex": {0: 0, 3: 2, 4: 3, 11: 6, 12: 3, 17: 3, 20: 1, 29: 5},
    "biored": {0: 0, 1: 3, 2: 3, 3: 2, 4: 5, 5: 4, 6: 6},
    "chemdner": {0: 0, 1: 2},
    "chemprot": {0: 0, 1: 2, 2: 3, 3: 3},
    "distemist": {0: 0},
    "ddi_corpus": {0: 0},
    "ctebmsp": {0: 0},
    "chia": {0: 0, 4: 3},
    "mlee": {
        0: 0,
        1: 3,
        2: 3,
        3: 6,
        4: 3,
        6: 5,
        7: 5,
        9: 3,
        10: 6,
        11: 3,
        14: 2,
        15: 5,
        16: 3,
    },
    "mirna": {0: 0, 2: 1, 6: 3},
    "nlmchem": {0: 0, 1: 2},
    "nlm_gene": {0: 0, 1: 3, 2: 3, 3: 3},
    "ncbi_disease": {0: 0, 1: 1, 2: 1},
    "pdr": {0: 0, 1: 1},
    "osiris": {0: 0, 1: 3, 2: 4},
    "progene": {0: 0, 1: 3},
    "pico_extraction": {0: 0},
    "tmvar_v2": {0: 0},
    "tmvar_v1": {0: 0},
    "spl_adr_200db": {0: 0, 3: 3},
    "seth_corpus": {0: 0, 1: 3},  # "seth_corpus": {0: 0, 1: 3, 2: 4},
    "scai_disease": {0: 0, 1: 1},
    "scai_chemical": {0: 0},
    "BC5CDR_TrainDev": {0: 0, 1: 2, 2: 1},
    "BioID_Train_revision": {0: 0, 1: 6},
    "DrugProt_Train": {0: 0, 1: 2, 2: 3},
    "GNormPlus_Train": {0: 0, 2: 3, 3: 3},
    "Linnaeus_Train": {0: 0, 1: 5},
    "NCBIdisease_TrainDev": {0: 0, 1: 1, 2: 1, 3: 1, 4: 1},
    "NLMchem_TrainDev": {0: 0, 1: 2},
    "NLMgene_Train": {0: 0, 1: 3, 2: 3},
    "SPECIES_Train": {0: 0, 1: 5},
    "tmVar3_Train": {
        0: 0,
        1: 3,
        2: 5,
        3: 1,
        4: 4,
        5: 4,
        6: 4,
        7: 6,
        8: 4,
        9: 4,
        10: 4,
        11: 4,
    },
}

# abst
abst_sentences = {
    "bc5cdr": "Manually curating chemicals, diseases, and their relations is of sig- nificant importance to biomedical research but is plagued by its high cost and the rapid growth of the biomedical literature. In recent years, there has been a growing interest to develop computational approaches for automatic chemical- disease relation (CDR) extraction with proposals of different techniques. De- spite these attempts, the lack of a comprehensive benchmarking dataset has lim- ited the comparison of different techniques in order to assess and advance the current state of the art. To this end, we set up a challenge task through BioCrea- tive V to automatically extract CDRs from the literature. More specifically, we designed two challenge tasks: disease named entity recognition (DNER) and chemical-induced disease (CID) relation extraction. To assist system develop- ment and assessment, we created a large annotated text corpus that consists of human annotations of all chemicals, diseases and their interactions in 1,500 PubMed articles. A total of 34 teams worldwide participated in the CDR task: 16 in the DNER task and 18 in the CID task. When comparing the text-mined results with the manually annotated ground truth, the best systems achieved an F-score of 86.46 for the DNER task – a result that approaches the human inter- annotator agreement (0.8875) – and an F-score of 57.03 for the CID task, the highest results ever reported for such tasks. In addition to the accuracy, another novel aspect of our evaluation is that we tested each participating system's abil- ity to return real-time results in a timely manner: the average response time for each team’s DNER and CID systems are 5.6 and 9.3 seconds via their respec- tive web services. Given the level of participant and team results, we find our task to be successful in engaging the text-mining research community, produc- ing a large annotated corpus, and improving the results of automatic disease recognition and chemical-disease relation extraction.' # 'BC5CDR corpus consists of 1500 PubMed articles with 5203 annotated chemicals, 4182 diseases and 3116 chemical-disease interactions.",
    # 'bc5cdr': "Community-run formal evaluations and manually annotated text cor- pora are critically important for advancing biomedical text mining research. Re- cently in BioCreative V, a new challenge was organized for the tasks of disease named entity recognition (DNER) and chemical-induced disease (CID) relation extraction. Given the nature of both tasks, a test collection is required to contain both disease/chemical annotations and relation annotations in the same set of ar- ticles. Despite previous efforts in biomedical corpus construction, none were found to be sufficient for the task. Thus, we developed our own corpus during the challenge by inviting a team of expert annotators from both MeSH and the Comparative Toxicogenomics Database (CTD), who performed manual annota- tion of entities (diseases/chemicals) and relations, respectively. To ensure high annotation quality and productivity, detailed annotation guidelines and automat- ic annotation tools were provided. The resulting corpus consists of 1,500 Pub- Med articles with 4,409 annotated chemicals, 5,818 diseases, and 3,116 chemi- cal-disease interactions. Each annotation includes both the mention text spans and normalized concept identifiers (MeSH was used as the controlled vocabu- lary). To ensure accuracy, the entities were captured independently by two an- notators; the average inter-annotator agreement (IAA) scores are 88.75% and 96.31% for the disease and chemicals, respectively, in the test set according to the Jaccard similarity coefficient. Our corpus was successfully used for the Bi- oCreative V challenge tasks and should serve as a valuable resource for the text-mining research community.",
    "ncbi": "Information encoded in natural language in biomedical literature publications is only useful if efficient and reliable ways of accessing and analyzing that information are available. Natural language processing and text mining tools are therefore essential for extracting valuable information, however, the development of powerful, highly effective tools to automatically detect central biomedical concepts such as diseases is conditional on the availability of annotated corpora. This paper presents the disease name and concept annotations of the NCBI disease corpus, a collection of 793 PubMed abstracts fully annotated at the mention and concept level to serve as a research resource for the biomedical natural language processing community. Each PubMed abstract was manually annotated by two annotators with disease mentions and their corresponding concepts in Medical Subject Headings (MeSH®) or Online Mendelian Inheritance in Man (OMIM®). Manual curation was performed using PubTator, which allowed the use of pre-annotations as a pre-step to manual annotations. Fourteen annotators were randomly paired and differing annotations were discussed for reaching a consensus in two annotation phases. In this setting, a high inter-annotator agreement was observed. Finally, all results were checked against annotations of the rest of the corpus to assure corpus-wide consistency. The public release of the NCBI disease corpus contains 6892 disease mentions, which are mapped to 790 unique disease concepts. Of these, 88% link to a MeSH identifier, while the rest contain an OMIM identifier. We were able to link 91% of the mentions to a single disease concept, while the rest are described as a combination of concepts. In order to help researchers use the corpus to design and test disease identification methods, we have prepared the corpus as training, testing and development sets. To demonstrate its utility, we conducted a benchmarking experiment where we compared three different knowledge-based disease normalization methods with a best performance in F-measure of 63.7%. These results show that the NCBI disease corpus has the potential to significantly improve the state-of-the-art in disease name recognition and normalization research, by providing a high-quality gold standard thus enabling the development of machine-learning based approaches for such tasks.",
}

# conc
conc_sentences = {
    "bc5cdr": "We developed a corpus for both named entity recognition and chemical-disease relations in the literature. A total of 1,500 articles have been annotated with automated assistance from PubTator. Jaccard agreement results and corpus statistics verified the reliability of the corpus. Furthermore, our annotated data includes the CDR relations that are asserted across sentence boundaries (i.e. not in the same sentences). We believe this data set will be invaluable for advancing text-mining techniques for relation extraction tasks.",
    "ncbi": "This work presents the NCBI disease corpus, a richly annotated corpus with disease names and their corresponding MeSH and/or OMIM identifiers. This resource contains 793 PubMed abstracts, and lists 6892 disease mentions, which are linked to 790 unique concepts, thus providing an important foundation for improving the text-mining research on disease named entity recognition. Our experiments demonstrated the feasibility of using the corpus as the basis for training learning models in both named entity identification and concept recognition, and we expect these results to serve as benchmark for other future methods. To facilitate future benchmarking experiments, the corpus is also divided into training, development and testing sets. The NCBI disease corpus contains annotations for all sentences in a PubMed document (title and abstract), an important aspect in facilitating development of complex information retrieval tasks that connect diseases to treatments, causes or other types of information. This corpus provides annotation of disease mentions in four major categories: Specific Disease (i.e., clearcell renal cell carcinoma), Disease Class (i.e., cystic kidney diseases), Composite mentions (i.e., prostatic, pancreas, skin, and lung cancer), and Modifier (i.e., hereditary breast cancer families) [21]. Disease normalization annotation guidelines were designed with the goal of allowing flexible matching to MeSH (diseases branch) and OMIM concepts, while retaining the true meaning of the specific mention. The current corpus was reviewed several times by several annotators and describes a refined scale of the annotation categories. The NCBI corpus can act as a basis for the development of more accurate machine learning systems for disease name recognition and normalization, as well as boost research into other areas of biomedical knowledge discovery pertaining to diseases.",
}

# ano
annotation_sentences = {
    "bc5cdr": "The task organizers followed the usual practice of biomedical corpus annotation for entity annotation: the annotators were asked to follow an initial set of guidelines when annotating the first 100 sample articles. Annotation discrepancies and questions were discussed and settled by the senior annotator; the annotation guidelines were revised accordingly. Detailed guidelines are available on the task website. For relation annotation, the CTD standard curation protocol was followed.",
    "ncbi": "For disease concept annotation, annotators were encouraged to use their domain knowledge, as well as any other public resources such as UMLS and Wikipedia. Initially, a set of 20 randomly chosen PubMed abstracts was used as a practice set for the development of annotation guidelines. After each annotator worked individually on the set, the results were shared and discussed among all annotators. The final annotation guidelines are summarized below and also made available at the corpus download website. To explain our annotation rules, we use the disease mentions in the following text (bolded) as a running example: ‘‘The SchwartzJampel syndrome (SJS) is a recessively inherited condition defined by myotonia, short stature, and bone dysplasia.’’",
}

# ano
annotation_guidlines = {
    "bc5cdr": "Annotate the most specific disease mentions and select the best-matching MeSH ID. For instance, the complete phrase “partial seizures” should be preferred over “seizures” as it is more specific. [SEP] Annotate minimum necessary text spans for a disease. For example, select “hypertension” instead of “sustained hypertension.” [SEP] Annotate all mentions of a disease entity in an abstract. All occurrences of the same disease mention should be marked, including duplicates within the same sentence. PubTator provides an automatic function to help mark the duplicate mentions via a single click. [SEP] Annotate mentions with morphological variations such as adjectives. For instance, “hypertensive” is annotated as “hypertension.” [SEP] Annotate abbreviations. Abbreviations should be annotated separately. For instance, “Huntington disease (HD)” should be separated into two annotations: “Huntington disease” and “HD” with the same concept id.",
    "ncbi": "Annotate the concept that matches the preferred name. There are several matches for ‘‘myotonia’’ in MEDIC: Myotonia [MESH:D009222], Myotonia Congenita [MESH:D009224], Myotonic Disorders [MESH:D020967], Myotonic Dystrophy [MESH:D009223], etc. For the mention ‘‘myotonia’’, the entry MESH:D009222 is selected because it provides a perfect match to this concept’s preferred name. Moreover, the descriptions of the other pos- sible matches do not provide a better ﬁt when taking into account the context of the article. [SEP] Annotate the concept that matches the synonym name, unless there is another concept that matches the preferred name. Several MEDIC entries may match the mention ‘‘bone dysplasia’’: Bone Diseases, Developmental [MESH:D001848], Gracile bone dysplasia [MESH:C537291], Fibrous Dysplasia of Bone [MESH:D005357], etc. For the mention ‘‘bone dysplasia’’, the entry MESH:D001848 is selected because it lists a synonym which is a perfect match to this mention.",
}

task_rel_labels = {
    "ace04": ["PER-SOC", "OTHER-AFF", "ART", "GPE-AFF", "EMP-ORG", "PHYS"],
    "ace05": ["ART", "ORG-AFF", "GEN-AFF", "PHYS", "PER-SOC", "PART-WHOLE"],
    "scierc": [
        "PART-OF",
        "USED-FOR",
        "FEATURE-OF",
        "CONJUNCTION",
        "EVALUATE-FOR",
        "HYPONYM-OF",
        "COMPARE",
    ],
}

color_list = [
    # 赤系
    "#FF0000",
    "#8B0000",
    "#FF6347",
    "#FA8072",
    # 青系
    "#0000FF",
    "#00008B",
    "#4169E1",
    "#87CEEB",
    # 緑系
    "#008000",
    "#006400",
    "#32CD32",
    "#90EE90",
    # 黄色系
    "#FFFF00",
    "#FFD700",
    "#FFA500",
    "#FFDAB9",
    # 紫系
    "#800080",
    "#8A2BE2",
    "#DA70D6",
    "#DDA0DD",
    # 茶色系
    "#A52A2A",
    "#D2691E",
    "#CD853F",
    "#DEB887",
    # シアン系
    "#00FFFF",
    "#008B8B",
    "#20B2AA",
    "#E0FFFF",
    # マゼンタ系
    "#FF00FF",
    "#8B008B",
    "#FF1493",
    "#FFB6C1",
    # 緑青系
    "#008080",
    "#2F4F4F",
    "#66CDAA",
    "#AFEEEE",
    # その他の色
    "#FF4500",
    "#9ACD32",
    "#FF69B4",
    "#4B0082",
    "#00FA9A",
    "#48D1CC",
    "#5F9EA0",
    "#F4A460",
    "#9370DB",
    "#3CB371",
    "#F0E68C",
    # グレー系
    "#000000",
    "#696969",
    "#A9A9A9",
    "#D3D3D3",
]

aioner_original = {
    "BC5CDR_TrainDev": 500,
    "BioID_Train_revision": 7763,
    "GNormPlus_Train": 281,
    "Linnaeus_Train": 64,
    "NCBIdisease_TrainDev": 593,
    "NLMchem_TrainDev": 5459,
    "NLMgene_Train": 360,
    "SPECIES_Train": 511,
    "tmVar3_Train": 274,
}


def get_labelmap(label_list, initial_value=0):
    label2id = {}
    id2label = {}
    for i, label in enumerate(label_list):
        label2id[label] = i + 1 + initial_value
        id2label[i + 1 + initial_value] = label
    return label2id, id2label
